src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs

```diff
@@ -1712,8 +2802,183 @@ class D
+            {
+                void M();
+            }
+
+            class C : I
+            {
+                [Interceptable] // 1
+                void I.M() => throw null!;
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 5, 8)]
+                public static void Interceptor(this I i) => throw null!;
+            }
+            """;
+
+        var comp = CreateCompilation(new[] { (source, "Program.cs"), s_attributesSource });
+        comp.VerifyEmitDiagnostics(
+            // Program.cs(14,6): error CS27008: An interceptable method must be an ordinary member method.
+            //     [Interceptable] // 1
+            Diagnostic(ErrorCode.ERR_InterceptableMethodMustBeOrdinary, "Interceptable").WithLocation(14, 6)
+            );
+    }
+
+    [Fact]
+    public void InterceptorExtern()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+
+            C.M();
+
+            class C
+            {
+                [Interceptable]
+                public static void M() => throw null!;
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 3, 3)]
+                public static extern void Interceptor();
+            }
+            """;
+
+        var comp = CreateCompilation(new[] { (source, "Program.cs"), s_attributesSource });
+        comp.VerifyEmitDiagnostics();
+    }
+
+    [Fact]
+    public void InterceptorAbstract()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            var d = new D();
+            d.M();
+
+            abstract class C
+            {
+                [Interceptable]
+                public void M() => throw null!;
+
+                [InterceptsLocation("Program.cs", 5, 3)]
+                public abstract void Interceptor();
+            }
+
+            class D : C
+            {
+                public override void Interceptor() => Console.Write(1);
+            }
+            """;
+
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, expectedOutput: "1");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void InterceptorInterface()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            I i = new C();
+            i.M();
+
+            interface I
+            {
+                [Interceptable]
+                public void M();
+
+                [InterceptsLocation("Program.cs", 5, 3)]
+                void Interceptor();
+            }
+
+            class C : I
+            {
+                public void M() => throw null!;
+                public void Interceptor() => Console.Write(1);
+            }
+            """;
+
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, expectedOutput: "1");
```

Comment:
Interesting problem! This is reporting errors non-deterministically.

```cs
Roslyn.Test.Utilities.EmitException : Emit Failed, binaries saved to:
C:\Users\rikki\AppData\Local\Temp\RoslynTests
Program.cs(12,6): error CS27016: The indicated call is intercepted multiple times.
Program.cs(12,6): error CS27016: The indicated call is intercepted multiple times.
```

Reply:
It's expected that attributes will be bound potentially multiple times concurrently on the same symbol. Therefore addressing this by handling some duplicate attribute detection in the interceptors map.